pkg/config/handlerConfigurer.go

```diff
@@ -0,0 +1,191 @@
+	// constructors based on template kind. This results into something like
+	// map[handlerName]map[TemplateName][][Constructor.InstanceName].
+	// 2. We then infer all the types for all the known
+	// constructors.
+	// 3. Using data from #1 and #2, for each handler and for each template within it, we call configure*TemplateName*
+	// with all the inferred types for all the instanceNames that belong to handler-template group.
+	configurer := handlerConfigurer{tmplRepo: tmplRepo, typeChecker: expr, attrDescFinder: df}
+
+	iTypes, err := configurer.inferTypes(constructors)
+	if err != nil {
+		return err
+	}
+	grpHandlers, err := configurer.groupHandlerInstancesByTemplate(actions, constructors, handlers)
+	if err != nil {
+		return err
+	}
+
+	return configurer.dispatchTypesToHandlers(iTypes, grpHandlers, handlers)
+}
+
+func (h *handlerConfigurer) dispatchTypesToHandlers(infrdTypes map[string]proto.Message,
+	instsByTmpls map[string]instancesByTemplate, handlers map[string]*HandlerBuilderInfo) error {
+	for hName, instsByTmpl := range instsByTmpls {
+		hb, found := handlers[hName]
+		if !found {
+			// This should not happen, since the dispatchTypesToHandlers should get called after all verifications.
+			panic(fmt.Errorf("handler %s is not registered. This code should be called after config has been "+
+				"validated", hName))
+		}
+
+		for tmplName, insts := range instsByTmpl.instancesNamesByTemplate {
+			ti, found := h.tmplRepo.GetTemplateInfo(tmplName)
+			if !found {
+				// This should not happen, since the dispatchTypesToHandlers should get called after all verifications.
+				panic(fmt.Errorf("template %s is not registered. This code should be called after config has "+
+					"been validated", tmplName))
+			}
+
+			typesToConfigure := make(map[string]proto.Message)
+			for _, inst := range insts {
+				v, found := infrdTypes[inst]
+				if !found {
+					// This should not happen, since the dispatchTypesToHandlers should get called after all
+					// verifications.
+					panic(fmt.Errorf("instance %s is not found in inferred types. This code should be called "+
+						"after config has been validated", inst))
+				}
+				typesToConfigure[inst] = v
+			}
+
+			// ConfigureTypeFn calls into handler's configure code which can panic. If that happens, we will
+			// remove the handler from the list of handlers to configure.
+			defer func() {
+				if r := recover(); r != nil {
+					glog.Warningf("handler '%s' panicked with '%v' when trying to configure it with instances "+
+						"%v. Please remove the handler or fix the configuration.", hName, r, insts)
+					hb.isBroken = true
+				}
+			}()
+
+			err := ti.ConfigureTypeFn(typesToConfigure, hb.handlerBuilder)
+			if err != nil {
+				glog.Warningf("Cannot configure handler %s with types %v: %v", hName, typesToConfigure, err)
+				return err
+			}
+		}
+	}
+	// TODO How to handle case where error in config/or adapter returns error, and we have done partial configuration.
```

Comment:
I don't think it matters: partially done = bad config = we should reject it and serve with our most recent valid config. Just throw away any incremental work done here.